state/compatibility_test.go

```diff
@@ -0,0 +1,225 @@
+	data, err := v1beta2ABCIResponsesWithNull.Marshal()
+	require.NoError(t, err)
+	require.NotNil(t, data)
+
+	// This should not work since they have different schemas
+	// but an error is not returned, so it deserializes an ABCIResponse
+	// on top of a FinalizeBlockResponse giving the false impression they are the same
+	// but because it doesn't error out, the fields in finalizeBlockResponse will have
+	// their zero-value (e.g. nil, 0, "")
+	finalizeBlockResponse := new(abci.FinalizeBlockResponse)
+	err = finalizeBlockResponse.Unmarshal(data)
+	require.NoError(t, err)
+	require.Nil(t, finalizeBlockResponse.AppHash)
+	require.Nil(t, finalizeBlockResponse.TxResults)
+}
+
+// This test unmarshal a v1beta2.ABCIResponses as a statev1beta3.LegacyABCIResponses
+// This logic is important for the LoadFinalizeBlockResponse method in the state store
+// The conversion should work because they are compatible.
+func TestStateV1Beta2ABCIResponsesAsStateV1Beta3FinalizeBlockResponse(t *testing.T) {
+	v1beta2ABCIResponses := newV1Beta2ABCIResponses()
+
+	data, err := v1beta2ABCIResponses.Marshal()
+	require.NoError(t, err)
+	require.NotNil(t, data)
+
+	// This works because they are equivalent protos and the fields are populated
+	legacyABCIResponses := new(statev1beta3.LegacyABCIResponses)
+	err = legacyABCIResponses.Unmarshal(data)
+	require.NoError(t, err)
+	require.NotNil(t, legacyABCIResponses.DeliverTxs)
+	require.NotNil(t, legacyABCIResponses.EndBlock)
+	require.NotNil(t, legacyABCIResponses.BeginBlock)
+}
+
+// This test unmarshal a v1beta2.ABCIResponses as a statev1beta3.LegacyABCIResponses
+// This logic is important for the LoadFinalizeBlockResponse method in the state store
+// The conversion should work because they are compatible even if fields to be converted are null.
+func TestStateV1Beta2ABCIResponsesWithNullAsStateV1Beta3FinalizeBlockResponse(t *testing.T) {
+	v1beta2ABCIResponsesWithNull := newV1Beta2ABCIResponsesWithNullFields()
+	data, err := v1beta2ABCIResponsesWithNull.Marshal()
+	require.NoError(t, err)
+	require.NotNil(t, data)
+
+	// This works because they are equivalent protos and the fields are populated
+	// even if a field is null, it will be converted properly
+	legacyResponseWithNull := new(statev1beta3.LegacyABCIResponses)
+	err = legacyResponseWithNull.Unmarshal(data)
+	require.NoError(t, err)
+	require.NotNil(t, legacyResponseWithNull.DeliverTxs)
+	require.Nil(t, legacyResponseWithNull.EndBlock)
+	require.NotNil(t, legacyResponseWithNull.BeginBlock)
+}
+
+// Generate a v1beta2 ABCIResponses with data for all fields.
+func newV1Beta2ABCIResponses() statev1beta2.ABCIResponses {
+	eventAttr := abciv1beta2.EventAttribute{
+		Key:   "key",
+		Value: "value",
+	}
+
+	deliverTxEvent := abciv1beta2.Event{
+		Type:       "deliver_tx_event",
+		Attributes: []abciv1beta2.EventAttribute{eventAttr},
+	}
+
+	responseDeliverTx := abciv1beta2.ResponseDeliverTx{
+		Code:   abci.CodeTypeOK,
```

Comment:
Should you populate the rest of the fields here? (`GasWanted`, `Codespace`, etc)